File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from itertools import combinations

def find_parent(x, parent):
    if parent[x] != x:
        parent[x] = find_parent(parent[x], parent)
    return parent[x]

def make_union(a, b, parent):
    a = find_parent(a, parent)
    b = find_parent(b, parent)
    if a < b:
        parent[b] = a
    else:
        parent[a] = b

def solution(n, wires):
    answer = float('inf')
    for w in combinations(wires, len(wires) - 1):
        parent = [i for i in range(n + 1)]
        for a, b in w:
            make_union(a, b, parent)
        
        root_counts = [0] * (n + 1)
        for i in range(1, n + 1):
            root = find_parent(i, parent)
            root_counts[root] += 1
        
        counts = [count for count in root_counts if count > 0]
        if len(counts) == 2:
            answer = min(answer, abs(counts[0] - counts[1]))
    
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

int find_parent(int x, vector<int>& parent) {
    if (parent[x] == x) return x;
    return parent[x] = find_parent(parent[x], parent); 
}

void make_union(int a, int b, vector<int>& parent) {
    a = find_parent(a, parent);
    b = find_parent(b, parent);
    if (a != b) {
        if (a < b) parent[b] = a;
        else parent[a] = b;
    }
}

int solution(int n, vector<vector<int>> wires) {
    int answer = n; 

    for (int i = 0; i < wires.size(); i++) {

        vector<int> parent(n + 1);
        for (int j = 1; j <= n; j++) parent[j] = j;

        for (int j = 0; j < wires.size(); j++) {
            if (i == j) continue; 
            make_union(wires[j][0], wires[j][1], parent);
        }

        int count = 0;
        int root = find_parent(1, parent);
        for (int j = 1; j <= n; j++) {
            if (find_parent(j, parent) == root) count++;
        }

        int diff = abs(count - (n - count));
        answer = min(answer, diff);
    }

    return answer;
}